resource/file/file.go

```diff
@@ -0,0 +1,566 @@
+// Copyright © 2016 Asteris, LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package file
+
+import (
+	"crypto/sha256"
+	"encoding/hex"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/user"
+	"path/filepath"
+	"strconv"
+
+	"github.com/asteris-llc/converge/resource"
+	"github.com/pkg/errors"
+)
+
+const defaultPermissions = os.FileMode(0750)
+
+// Action is the calculated action
+type Action uint8
+
+const (
+	// ActionNone indicates no action is required
+	ActionNone Action = iota
+
+	// ActionCreate creates a File resource
+	ActionCreate
+
+	// ActionModify deletes a File resource
+	ActionModify
+
+	// ActionDelete deletes a File resource
+	ActionDelete
+)
+
+func (a Action) String() string {
+	return string(a)
+}
+
+// State is the desired state of the file
+type State string
+
+const (
+	// StateAbsent indicates the file should be absent
+	StateAbsent State = "absent"
+
+	// StateUndefined indicates an undefined state
+	StateUndefined State = ""
+
+	// StatePresent indicates the file should be present
+	StatePresent State = "present"
+
+	// DefaultState is the default state
+	DefaultState State = StatePresent
+)
+
+// ValidStates indicates supported states that can be defined
+var ValidStates = []State{StateAbsent, StatePresent}
+
+func (s State) String() string {
+	return string(s)
+}
+
+// Type is the type of file
+type Type string
+
+const (
+	// TypeDirectory is a directory
+	TypeDirectory Type = "directory"
+
+	// TypeFile is a regular file
+	TypeFile Type = "file"
+
+	// TypeLink is a hardlink
+	TypeLink Type = "hardlink"
+
+	// TypeNone is an undefined type
+	TypeNone Type = ""
+
+	// TypeSymlink is a symlink
+	TypeSymlink Type = "symlink"
+
+	// DefaultType is the default type
+	DefaultType Type = TypeFile
+)
+
+var (
+	// ValidFileTypes are valid file types
+	ValidFileTypes = []Type{TypeDirectory, TypeFile}
+
+	// ValidLinkTypes are valid link types
+	ValidLinkTypes = []Type{TypeLink, TypeSymlink}
+
+	// AllTypes are the current FileTypes that are supported
+	AllTypes = append(ValidFileTypes, ValidLinkTypes...)
+)
+
+func (t Type) String() string {
+	return string(t)
+}
+
+// File contains information for managing files
+type File struct {
+	Destination string
+	State       State
+	Type        Type
+	Target      string
+	Force       bool    //force replacement of file
+	Mode        *uint32 //requested permissions from Prepare
```

Comment:
this probably should be a non-pointer FileMode type by the time it comes from Prepare